File: scripts-python/scraper/goodrx_scraper.py
from playwright.sync_api import sync_playwright
import requests
import json

WORKER_URL = "https://transparentrx-pricing.kellybhorak.workers.dev/api/retail-price"

drug = "lisinopril"
dose = "20mg"
quantity = 30
zip_code = "76102"

url = f"https://www.goodrx.com/{drug}?dosage={dose}&form=tablet&quantity={quantity}&label_override={drug}"

def send_price(data):
    try:
        r = requests.post(WORKER_URL, json=data)
        print("sent:", r.status_code)
    except Exception as e:
        print("send error:", e)

with sync_playwright() as p:

    browser = p.chromium.launch(headless=True)
    page = browser.new_page()

    print("Loading page...")
    page.goto(url)

    page.wait_for_timeout(6000)

    pharmacies = page.evaluate("""
    () => {
        const rows = document.querySelectorAll('[data-test="pharmacy-row"]')
        const data = []

        rows.forEach(row => {

            const name = row.querySelector('[data-test="pharmacy-name"]')?.innerText
            const price = row.querySelector('[data-test="price"]')?.innerText

            if(name && price){
                data.push({
                    pharmacy:name,
                    price:price.replace(/[^0-9\\.]/g,'')
                })
            }

        })

        return data
    }
    """)

    browser.close()

for p in pharmacies:

    payload = {
        "ndc": "00054018113",
        "drug_name": drug,
        "strength": dose,
        "quantity": quantity,
        "pharmacy_name": p["pharmacy"],
        "pharmacy_chain": p["pharmacy"],
        "cash_price": float(p["price"]),
        "coupon_price": None,
        "zip_code": zip_code,
        "source": "goodrx"
    }

    print(payload)

    send_price(payload)

File: goodrx-playwright/goodrx_scraper.js
require("dotenv").config()
const { chromium } = require("playwright")

async function scrapeGoodrx(drug, strength, quantity, zip) {

  const browser = await chromium.launch({ headless: false })
  const page = await browser.newPage()

  const url =
    `https://www.goodrx.com/${drug}?dosage=${strength}&form=tablet&quantity=${quantity}&label_override=${drug}&zip_code=${zip}`

  console.log("Opening:", url)

  // safer navigation mode
  await page.goto(url, { waitUntil: "domcontentloaded", timeout: 60000 })

  // allow React components to render
  await page.waitForTimeout(5000)

  const prices = await page.evaluate(() => {

    const results = []

    const pharmacies = [
      "CVS",
      "Walgreens",
      "Walmart",
      "Costco",
      "Kroger",
      "Rite Aid"
    ]

    const elements = document.querySelectorAll("div")

    elements.forEach(el => {

      const text = el.innerText || ""

      const pharmacyMatch = pharmacies.find(p =>
        text.toLowerCase().includes(p.toLowerCase())
      )

      const priceMatch = text.match(/\$[0-9]+\.[0-9]+/)

      if (pharmacyMatch && priceMatch) {

        results.push({
          pharmacy: pharmacyMatch,
          price: priceMatch[0]
        })

      }

    })

    return results

  })

  await browser.close()

  return prices

}

async function run() {

  const results = await scrapeGoodrx(
    "lisinopril",
    "10mg",
    30,
    process.env.LOCATION
  )

  console.log(results)

}

run()
